Disable Cancel and Submit Buttons in Project Editor

JoshuaSchaefer opened the pull request:
---

Disable the cancel and submit buttons in the project editor when there are no changes to undo or save.

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroweb-Symfony/blob/develop/.github/contributing.md) and [wiki pages](https://github.com/Catrobat/catroweb-Symfony/wiki/) of this repository.

- [ ] Include the name and id of the Jira ticket in the PR’s title eg.: `SHARE-666 The devils ticket`
- [ ] Choose the proper base branch (*develop*)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no warnings and errors 
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Verify that all tests are passing (CI), if not please state the test cases in the [section](#Tests) below
- [ ] Perform a self-review of the changes
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] Put your ticket into the `Code Review` section in [Jira](https://jira.catrob.at/)
- [ ] Post a message in the *#catroweb* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer
- [ ] Check that your pull request has been successfully deployed to https://web-test-1.catrob.at/

### Additional Description
Additional information on the custom translation feature can be found here: [catrobat custom translations](https://docs.google.com/document/d/1IU-12Np7yajbz2Gt9rkfm2jasQlt7trhliu1gqE1k4Y/edit?usp=sharing)

### Tests - additional information
Added new checks to make sure that cancel and save buttons are disabled and enabled in the correct situation.

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -62,6 +64,8 @@ export function ProjectEditor (projectDescriptionCredits, programId, textFields)
     }
   })
 
+  document.addEventListener('keyup', checkTextFields)
```
Is there a more specific listener on each text fields? I thought there was a `lastSavedText` field that could be useful?

JoshuaSchaefer replied:
The lastSavedText holds the last saved value for the text field so it is really only useful to check if the current text is the same as the last saved text.

I tried to see if there were any other handlers for the text fields that would trigger when the text field was changed, but I couldn't really find anything more straightforward than the 'keyup' event listener. It seems to work pretty well for what we want right now.

danielzhang130 replied:
have you tried this? https://stackoverflow.com/questions/6169353/html-text-input-onchange-event
i'm a bit concerned with the listener on the whole document, instead of the specific text fields

danielzhang130 replied:
Also we can add a small delay with debouncing so things don't react so often
https://stackoverflow.com/questions/27787768/debounce-function-in-jquery

JoshuaSchaefer replied:
Updated it to a .change listener which is only called when the text field is changed and it is unselected so it does not get called as nearly as often

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -15,6 +15,8 @@ export function ProjectEditor (projectDescriptionCredits, programId, textFields)
   this.selectedLanguage = $('#edit-selected-language')
   this.textLoadingSpinner = $('#edit-loading-spinner')
+  this.saveButton = $('#edit-submit-button')
+  this.cancelButton = $('#edit-cancel-button')
```
This could now be replaced by the new local variable

https://github.com/danielzhang130/Catroweb/blob/fd3f1dcf5ebd96a94c919bfee062d21292439b93/assets/js/components/ProjectEditor.js#L48

JoshuaSchaefer replied:
Should be updated now

danielzhang130 reviewed (commented):
I feel like since this change is small, we can take the opportunity to optimize the tests a bit. I think adding additional steps to these scenarios will make them even more flaky. Maybe there could be just one scenario that tests the enabling, disabling of buttons, since it's implemented all by the same class `ProjectEditor`. Although you could argue that we need to test enabling, disabling of buttons in all scenarios since tests would not be dependent on the implementation choice.... idk...

JoshuaSchaefer commented:
> I feel like since this change is small, we can take the opportunity to optimize the tests a bit. I think adding additional steps to these scenarios will make them even more flaky. Maybe there could be just one scenario that tests the enabling, disabling of buttons, since it's implemented all by the same class `ProjectEditor`. Although you could argue that we need to test enabling, disabling of buttons in all scenarios since tests would not be dependent on the implementation choice.... idk...

Yeah, I was on the fence about this as well. I think to completely test all the scenarios there would be a lot of repeated tests which is why I ended up just adding the steps to the existing tests. I think it is okay to just add the steps to the existing tests since they are pretty straightforward checks. If they become really flaky with the extra steps then I could probably remove the checks in a few tests that are somewhat redundant and just add a couple of simple tests to make sure it is behaving properly. For now, I think it is okay and we can just monitor it.

danielzhang130 commented:
> > I feel like since this change is small, we can take the opportunity to optimize the tests a bit. I think adding additional steps to these scenarios will make them even more flaky. Maybe there could be just one scenario that tests the enabling, disabling of buttons, since it's implemented all by the same class `ProjectEditor`. Although you could argue that we need to test enabling, disabling of buttons in all scenarios since tests would not be dependent on the implementation choice.... idk...
> 
> Yeah, I was on the fence about this as well. I think to completely test all the scenarios there would be a lot of repeated tests which is why I ended up just adding the steps to the existing tests. I think it is okay to just add the steps to the existing tests since they are pretty straightforward checks. If they become really flaky with the extra steps then I could probably remove the checks in a few tests that are somewhat redundant and just add a couple of simple tests to make sure it is behaving properly. For now, I think it is okay and we can just monitor it.

The tests are already flaky without them...

JoshuaSchaefer commented:
> > > I feel like since this change is small, we can take the opportunity to optimize the tests a bit. I think adding additional steps to these scenarios will make them even more flaky. Maybe there could be just one scenario that tests the enabling, disabling of buttons, since it's implemented all by the same class `ProjectEditor`. Although you could argue that we need to test enabling, disabling of buttons in all scenarios since tests would not be dependent on the implementation choice.... idk...
> > 
> > 
> > Yeah, I was on the fence about this as well. I think to completely test all the scenarios there would be a lot of repeated tests which is why I ended up just adding the steps to the existing tests. I think it is okay to just add the steps to the existing tests since they are pretty straightforward checks. If they become really flaky with the extra steps then I could probably remove the checks in a few tests that are somewhat redundant and just add a couple of simple tests to make sure it is behaving properly. For now, I think it is okay and we can just monitor it.
> 
> The tests are already flaky without them...

Honestly, I am not exactly sure how to fix them and it doesn't look like the added steps are causing them to fail at a higher rate from what I have seen.

danielzhang130 commented on assets/js/components/ProjectEditor.js:
```diff
@@ -43,7 +45,7 @@ export function ProjectEditor (projectDescriptionCredits, programId, textFields)
-  $('#edit-submit-button').on('click', () => { this.save() })
+  this.saveButton.on('click', () => { this.save() })
 
   $('#edit-cancel-button').on('click', () => { this.cancelChanges() })
```
this as well

JoshuaSchaefer replied:
Forgot that one, it is updated now

danielzhang130 commented:
while we are at it, should the cancel button be btn-secondary?

JoshuaSchaefer commented:
> while we are at it, should the cancel button be btn-secondary?

In the design they gave us the cancel button is btn-primary so I think we should just keep it the same as the design

Commit:
Disable Cancel and Submit Buttons in Project Editor

Disable the cancel and submit buttons in the project editor when there are no changes to undo or save.
